go/oasis-test-runner/scenario/e2e/runtime.go

```diff
@@ -607,34 +607,53 @@ func (sc *runtimeImpl) initialEpochTransitions() error {
 	if len(sc.net.Keymanagers()) > 0 {
 		// First wait for validator and key manager nodes to register. Then perform an epoch
 		// transition which will cause the compute and storage nodes to register.
-		numNodes := len(sc.net.Validators()) + len(sc.net.Keymanagers())
-		sc.logger.Info("waiting for (some) nodes to register",
-			"num_nodes", numNodes,
+		sc.logger.Info("waiting for validators to initialize",
+			"num_validators", len(sc.net.Validators()),
 		)
-
-		// TODO: once #2130 is done, aditionally wait for nodes to be Ready here.
-		// As generally a keymanager can register (with missing extra info)
-		// before actually being initialized and triggering an epoch transition
-		// at that point would be too soon.
-
-		if err := sc.net.Controller().WaitNodesRegistered(ctx, numNodes); err != nil {
-			return fmt.Errorf("failed to wait for nodes: %w", err)
+		for _, n := range sc.net.Validators() {
+			if err := n.WaitReady(ctx); err != nil {
+				return fmt.Errorf("failed to wait for a validator: %w", err)
+			}
+		}
+		sc.logger.Info("waiting for key managers to initialize",
+			"num_keymanagers", len(sc.net.Keymanagers()),
+		)
+		for _, n := range sc.net.Keymanagers() {
+			if err := n.WaitReady(ctx); err != nil {
+				return fmt.Errorf("failed to wait for a key manager: %w", err)
+			}
 		}
-
 		sc.logger.Info("triggering epoch transition")
 		if err := sc.net.Controller().SetEpoch(ctx, 1); err != nil {
 			return fmt.Errorf("failed to set epoch: %w", err)
 		}
 		sc.logger.Info("epoch transition done")
 	}
 
-	// Wait for all nodes to register.
-	sc.logger.Info("waiting for (all) nodes to register",
-		"num_nodes", sc.net.NumRegisterNodes(),
+	// Wait for storage workers, compute workers, and byzantine nodes to become ready.
+	sc.logger.Info("waiting for storage workers to initialize",
+		"num_storage_workers", len(sc.net.StorageWorkers()),
 	)
-
-	if err := sc.net.Controller().WaitNodesRegistered(ctx, sc.net.NumRegisterNodes()); err != nil {
-		return fmt.Errorf("failed to wait for nodes: %w", err)
+	for _, n := range sc.net.StorageWorkers() {
+		if err := n.WaitReady(ctx); err != nil {
+			return fmt.Errorf("failed to wait for a storage worker: %w", err)
+		}
+	}
+	sc.logger.Info("waiting for compute workers to initialize",
+		"num_compute_workers", len(sc.net.ComputeWorkers()),
+	)
+	for _, n := range sc.net.ComputeWorkers() {
+		if err := n.WaitReady(ctx); err != nil {
+			return fmt.Errorf("failed to wait for a compute worker: %w", err)
+		}
+	}
+	sc.logger.Info("waiting for byzantine nodes to initialize",
+		"num_byzantine", len(sc.net.Byzantine()),
+	)
+	for _, n := range sc.net.Byzantine() {
+		if err := n.WaitReady(ctx); err != nil {
```

Comment:
I don't think this will work for Byzantine nodes? Byzantine nodes do not use the regular node APIs? Feel free to keep using `WaitNodesRegistered` here instead of adding support for `WaitReady` to Byzantine nodes (which doesn't make sense).